### PR TITLE
Rework pre-built dist upload/download

### PR DESCRIPTION
**Pull request by martinpitt**
In publish-dist, directly commit dist/ and related files to the -dist
repo (the ones which make_dist actually needs), and tag the commit with
the SHA, so that they are still easy to look up.  Drop the "old tarball"
removal. This uses git in a much more efficient way, as it now only
stores the deltas. In particular, for commits/PRs which don't change
src/, that will just add an additional tag to the previous build and
store no extra files at all. This avoids regularly overflowing the -dist
repo and making `git push` hang.

Simplify build-dist to do away with the intermediate tarball, as there
are not so many files to make upload-artifact inefficient.

Push every build as an individual tag only, without having them on a
branch. That avoids the need for force-pushing and history rewriting,
which fixes the race of publish-dist against prune-dist.  There is also
no more race condition between parallel publish-dist workflows.

Change prune-dist to keep the tags of the last 7 days plus current
master HEAD, and let garbage collection take care of cleaning up older
ones.

Rework make_dist to get the dist repo in ~/.cache/cockpit-dev/, so that
it persists across project checkouts, and shallow-fetch the desired SHA
build. Unpacking requires some care: Unfortunately our dist tarball
generation does not deal with sub-second file mtime differences, which
can invert the relative mtime of package-lock.json vs.
dist/*/manifest.json; to avoid this, only touch the webpack-built files,
but leave the `npm install` ones alone.

Taken from cockpit-machines:
https://github.com/cockpit-project/cockpit-machines/commit/612bb6aa8d1f
https://github.com/cockpit-project/cockpit-machines/commit/ae57e6249397

-----

I tested this on my fork. The [dist repo](https://github.com/martinpitt/cockpit-dist) now has two tags, and otherwise just an empty master branch. [build-dist](https://github.com/martinpitt/cockpit/actions/runs/784970139) and [publish-dist](https://github.com/martinpitt/cockpit/actions/runs/784999140) look as expected. I also ran [prune-dist](https://github.com/martinpitt/cockpit/runs/2436773139?check_suite_focus=true) -- of course there are not yet any tags older than a week, but it shows that it correctly recognized current master HEAD and the two tags.

The difference to the cockpit-machines workflows is now minimal. I will send a follow-up workflow to c-machines to further reduce it.

Downloading also works fine. I tested this with:

    git clean -ffdx; ln -s ../bots; GITHUB_BASE=martinpitt/cockpit test/image-prepare -qv fedora-34

```
[...]
From https://github.com/martinpitt/cockpit-dist
 * tag               sha-8b214e16d7ea3088131209567212d7aa29af5f59 -> FETCH_HEAD
make_dist: Extracting cached node_modules...
make_dist: Extracting cached package-lock.json...
make_dist: Extracting cached dist...
make_dist: Extracting cached tools/debian/copyright...
```
This uses the cached download, and then making the tarball is rather fast (~ 10 s).

### CODE REVIEW

**Comment by martinpitt** [comment]
Note that this will slightly regress the time `make_dist` takes, as the extra autogen.sh/make dist run now takes an additional 14 seconds. But the current "store full tarballs in git" approach is not viable, I don't want to regularly recreate the -dist repo because git push times out.

But that's still much better than the ~ 10 minutes that npm install and webpack would take locally. As a little consolation, downloading the cache should now generally be faster than before.

**Comment by martinpitt** [comment]
packit can't succeed here, as the publish-dist changes go together with the build-dist changes, and the former only runs from master.

File: .github/workflows/publish-dist.yml
Comment (allisonkarlitskaya) [inline .github/workflows/publish-dist.yml:31]: This is insecure.

The SHA came from a workflow which is under complete control of a potentially-hostile pull request.  Someone could open a pull request against cockpit, populate our current master branch SHA into this file, and then we'd pick it up here and tag their malicious payload as a valid cache of the current state of our master branch.

This is what I've been spending my week worrying about... I think we need to do better here.

One approach might be to upload things to git only after stuff lands on master, and use some form of artifact-based downloading for testing pull requests...

Another approach might be to use `pull_request_target` and try to come up with a way to harden the build.  For the `node_modules` stuff I'm running in a separate docker container, for example...

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: .github/workflows/publish-dist.yml
Comment (martinpitt) [reply .github/workflows/publish-dist.yml:31]: Of course it is untrusted, that's what the current master workflow has as well. But that's fine -- us project mainainers must review a PR. It's just a cache, it's exactly the same as locally checking out and building an unreviewed PR.

Comment (martinpitt) [reply .github/workflows/publish-dist.yml:31]: Discussed in meeting. Will rework this, in a separate PR. Either with:
 * Not storing the SHA in the workflow, but instead querying the github.event.workflow for the originally proposed repo/branch and determining the SHA that way.
 * If that somehow does not work out, restrict publish-dist to PRs whose owners are in cockpit-project. We'd lose packit tests for "outsider" PRs, but that does not hurt that much. (At least until we figure out a better way)

File: .github/workflows/prune-dist.yml
Comment (allisonkarlitskaya) [inline .github/workflows/prune-dist.yml:14]: Is this just to get `git rev-parse HEAD` below?  I feel like you could get that a lot easier than cloning everything...

File: .github/workflows/publish-dist.yml
Comment (allisonkarlitskaya) [inline .github/workflows/publish-dist.yml:14]: I'm really going to insist on this point: please pin this down to a specific sha, at the very least.

File: .github/workflows/publish-dist.yml
Comment (allisonkarlitskaya) [inline .github/workflows/publish-dist.yml:41]: Usual quoting complaint here:  `$tag` is coming from `$SHA` which is fed to us from an untrusted workflow... this could do some serious damage.

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:79]: It probably makes more sense to use `--git-dir` instead of `-C`.  Also: consider passing the directory as an argument to `init` rather.

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:74]: I'd expect `+ ".git"` here.

Also: i feel like the entire thing would read a lot easier as an f-string.  `os.path.join()` is a bit silly in places like this and really harms readability.

Also, very very minor gripe: I hate that we call `expanduser` even if `XDG_CACHE_HOME` is set...

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:106]: `FETCH_HEAD` is unsafe here: you might be building more than one thing on the same system, sharing the cache directory.  Use the tag.  Also, more `--git-dir`.

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:97]: I feel like with the order that we unpack things, we could probably *always* touch.  That would be a nice cleanup to some otherwise kinda-evil code.

Even if not: please just write this as a normal conditional inside the loop.

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:108]: almost feels like using `check_call` with `shell=True` might be easier.  Your call.

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:109]: Does python guarantee that the expression after an `assert` is executed?  We need this to reap the zombie...

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: .github/workflows/build-dist.yml
Comment (martinpitt) [reply .github/workflows/build-dist.yml:23]: Unfortunately no, see commit 8917ad9c77a1c6016d1463785af96c5dfd3a1c67

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py:97]: I did start out with "always touch", but that really gets in the way. `make` is not good with files that have sub-second resolution time differences, this leads to strange and not always 100% reliably reproducible failures, at least on my system. This does not happen during a "real" build, as building a webpack or npm install take way more than one second. The only alternative that I found was to add `time.sleep(1)` in between, but I found that way more ugly.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py:108]: I really don't like running dynamically built shell commands, especially not if they are constructed from user-settable variables such as `$GITHUB_BASE`, or may have paths with spaces.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py:106]: I had some trouble with the tag, as `git fetch origin <tagname>` does not quite fetch the actual tag, you get "not a valid object name: sha-ba9...". Indeed refs/tags/ is empty. And `git fetch --tag origin <tagname>` fetches *all* tags. Your PR #15700 has exactly the same problem. I'll do some googling/experimentation.

File: .github/workflows/publish-dist.yml
Comment (martinpitt) [reply .github/workflows/publish-dist.yml]: Fixed in separate commit, similar to https://github.com/cockpit-project/cockpit-machines/pull/143

File: .github/workflows/prune-dist.yml
Comment (martinpitt) [reply .github/workflows/prune-dist.yml]: Indeed. Replaced with `git ls-remote`.

File: .github/workflows/publish-dist.yml
Comment (martinpitt) [reply .github/workflows/publish-dist.yml]: Fixed.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py]: Thanks! Done.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py]: Added .git. I kept os.path.join, I prefer explicit path components over the more brittle string interpolation, but I agree that this is just a personal whim.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py]: I'm not sure -- I split it, to be on the safe side.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py]: I replaced the fetching with the magic incantation to *actually* fetch the tag, now it works fine. I also followed up to your PR to do the same there.

**Comment by martinpitt** [comment]
Re-tested on my fork: [publish-dist succeeded](https://github.com/martinpitt/cockpit/runs/2490029929?check_suite_focus=true), it [pushed the new tag correctly](https://github.com/martinpitt/cockpit-dist/tags), [prune-dist](https://github.com/martinpitt/cockpit/runs/2490046703?check_suite_focus=true) correctly recognized current master HEAD, and `GITHUB_BASE=martinpitt/cockpit test/make_dist.py -d` downloads the cache and builds a tar in ~ 10 seconds.

File: test/make_dist.py
Comment (martinpitt) [reply test/make_dist.py:98]: BTW, the commit message does explain that.

Comment (martinpitt) [reply test/make_dist.py]: I also added a comment right here.

File: .github/workflows/publish-dist.yml
Comment (martinpitt) [reply .github/workflows/publish-dist.yml]: Done in a separate commit, like in cockpit-machines in https://github.com/cockpit-project/cockpit-machines/pull/143

File: test/make_dist.py
Comment (allisonkarlitskaya) [inline test/make_dist.py:86]: Didn't know about `tag <tag>`.  cool.

File: .github/workflows/publish-dist.yml
Comment (allisonkarlitskaya) [inline .github/workflows/publish-dist.yml:36]: :grinning: 

Same comment about single quotes as for c-machines, if you care.

**Review by allisonkarlitskaya — APPROVED** [review]
There are so many wins here...

Thanks for dealing with all the bickering :)

File: .github/workflows/publish-dist.yml
Comment (martinpitt) [reply .github/workflows/publish-dist.yml:36]: Yup, just sent https://github.com/cockpit-project/cockpit-machines/pull/145 to synchronize/cleanup the workflows.

**Comment by martinpitt** [comment]
I deleted and re-initialized https://github.com/cockpit-project/cockpit-dist , so that we get rid of all the old cruft.

**Pull request merged**
